### PR TITLE
arrange() only effective at end of pipe, window_order() overrides

### LINKED ISSUE #276: Rethink optimization of ORDER BY
`ORDER BY` clauses in subqueries are problematic:

- MSSQL outrights forbids them (#275)
- Other DBMS sometimes ignore them (see reprex) and are not obliged to follow either (https://mariadb.com/kb/en/library/why-is-order-by-in-a-from-subquery-ignored/, https://dba.stackexchange.com/a/83500/6934)

As a stop gap, we could remove `ORDER BY` clauses from inner queries if we don't combine the inner and the outer query. With a warning. This would also solve #275. I wonder if this generates spurious warnings, e.g., for window functions.

Alternatively, we could allow `arrange()` only at the very last step in the pipe, and transform it to `window_order()` if additional verbs are added, with a warning. This will be a problem later when we want to support lazy operations for data frames.

For a true solution, we cannot simply delay the computation due to aliasing (https://github.com/tidyverse/dbplyr/issues/94#issuecomment-395599454). A few examples, and a reprex illustrating the problem (using #277), are below. This might well be out of scope.

```r
# Input:
lazy_frame(a = 1:3, b = 4:2) %>%
  mutate(a = -a) %>%
  arrange(a) %>%
  mutate(a = -a)

# Equivalent with `arrange()` delayed:
lazy_frame(a = 1:3, b = 4:2) %>%
  mutate(a = -a) %>%
  mutate(zzz_001 = a) %>%
  mutate(a = -a) %>%
  arrange(zzz_001) %>%
  select(-zzz_001)

# Input:
lazy_frame(a = 1:3, b = 4:2) %>%
  arrange(a) %>%
  group_by(b) %>%
  mutate(c = cumsum(a)) %>%
  ungroup()

# Equivalent with `arrange()` delayed:
lazy_frame(a = 1:3, b = 4:2) %>%
  window_order(a) %>%
  group_by(b) %>%
  mutate(c = cumsum(a)) %>%
  ungroup() %>%
  arrange(a)
```

``` r
library(tidyverse)
devtools::load_all("~/git/R/dbplyr")
#> Loading dbplyr
#> 
#> Attaching package: 'testthat'
#> The following object is masked from 'package:dplyr':
#> 
#>     matches
#> The following object is masked from 'package:purrr':
#> 
#>     is_null
#> Registering testing src: df OK
#> Registering testing src: sqlite OK
#> Registering testing src: mysql OK
#> Registering testing src: MariaDB OK
#> Registering testing src: postgres OK
#> Registering testing src: MSSQL OK


test_frame(a = 1:3, b = 4:2) %>%
  map(
    . %>%
      mutate(a = -a) %>%
      arrange(a) %>%
      mutate(a = -a)
  )
#> Created a temporary table named: ##dbplyr_001
#> $df
#> # A tibble: 3 x 2
#>       a     b
#>   <int> <int>
#> 1     3     2
#> 2     2     3
#> 3     1     4
#> 
#> $sqlite
#> # Source:     lazy query [?? x 2]
#> # Database:   sqlite 3.25.3 [:memory:]
#> # Ordered by: a
#>       a     b
#>   <int> <int>
#> 1     3     2
#> 2     2     3
#> 3     1     4
#> 
#> $mysql
#> # Source:     lazy query [?? x 2]
#> # Database:   mysql 5.5.5-10.1.38-MariaDB-0ubuntu0.18.04.1
#> #   [kirill@localhost:/test]
#> # Ordered by: a
#>       a     b
#>   <dbl> <int>
#> 1     1     4
#> 2     2     3
#> 3     3     2
#> 
#> $MariaDB
#> # Source:     lazy query [?? x 2]
#> # Database:   mysql 5.5.5-10.1.38-MariaDB-0ubuntu0.18.04.1
#> #   [kirill@localhost:/test]
#> # Ordered by: a
#>   a                   b
#>   <S3: integer64> <int>
#> 1 1                   4
#> 2 2                   3
#> 3 3                   2
#> 
#> $postgres
#> # Source:     lazy query [?? x 2]
#> # Database:   postgres 9.6.9 [kirill@/var/run/postgresql:5432/kirill]
#> # Ordered by: a
#>       a     b
#>   <int> <int>
#> 1     3     2
#> 2     2     3
#> 3     1     4
#> 
#> $MSSQL
#> # Source:     lazy query [?? x 2]
#> # Database:   Microsoft SQL Server 12.00.1300[@cynkra-mssql/main]
#> # Ordered by: a
#>       a     b
#>   <int> <int>
#> 1     1     4
#> 2     2     3
#> 3     3     2
```

<sup>Created on 2019-04-09 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

### LINKED ISSUE #373: Retaining too optimistic order information when manipulating after arrange()
It seems that `op_sort()` is a tad too optimistic in some cases.

Related: #338 #337. I think `arrange()` should only have an effect if it's the last verb in a pipe (or can be bubbled down safely), or to define the window order for window functions.

``` r
library(tidyverse)
library(dbplyr)

tbl <- memdb_frame(a = 1:3, b = 3:1)
tbl
#> # Source:   table<dbplyr_001> [?? x 2]
#> # Database: sqlite 3.29.0 [:memory:]
#>       a     b
#>   <int> <int>
#> 1     1     3
#> 2     2     2
#> 3     3     1

tbl_a <- tbl %>% arrange(a)
tbl_a
#> # Source:     table<dbplyr_001> [?? x 2]
#> # Database:   sqlite 3.29.0 [:memory:]
#> # Ordered by: a
#>       a     b
#>   <int> <int>
#> 1     1     3
#> 2     2     2
#> 3     3     1

# Sort order wrong, the database won't guarantee lexical ordering
tbl_ab <- tbl %>% arrange(a) %>% arrange(b)
tbl_ab %>% sql_render()
#> <SQL> SELECT *
#> FROM (SELECT *
#> FROM `dbplyr_001`
#> ORDER BY `a`)
#> ORDER BY `b`

# a no longer exists, confusing
tbl_aba <- tbl %>% arrange(a) %>% select(-a)
tbl_aba
#> # Source:     lazy query [?? x 1]
#> # Database:   sqlite 3.29.0 [:memory:]
#> # Ordered by: a
#>       b
#>   <int>
#> 1     3
#> 2     2
#> 3     1

# How can I un-arrange?
tbl_aa <- tbl_aba %>% arrange(a) %>% arrange()
tbl_aa
#> # Source:     lazy query [?? x 1]
#> # Database:   sqlite 3.29.0 [:memory:]
#> # Ordered by: a, a
#>       b
#>   <int>
#> 1     3
#> 2     2
#> 3     1

# The ordering leaks to the lag() call
tbl_ac <- tbl_aba %>% collapse() %>% mutate(c = lag(b))
tbl_ac
#> Error: no such column: a
tbl_ac %>% sql_render()
#> <SQL> SELECT `b`, LAG(`b`, 1, NULL) OVER (ORDER BY `a`) AS `c`
#> FROM (SELECT `b`
#> FROM (SELECT *
#> FROM `dbplyr_001`
#> ORDER BY `a`))
```

<sup>Created on 2019-11-01 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by krlmlr**
- `arrange()` can now only be used at the end of a pipe or followed by `head()`, otherwise it has no effect and raises a warning.
- `window_order()` can be used repeatedly but only affects the verb that follows immediately.
- All other verbs reset the sort order, with warning.
- An empty `arrange()` can be used to undo the effects of the last `arrange()`.
- This change leads to fewer subqueries and fewer redundant `ORDER BY` clauses.
- The documentation for `arrange()` and `window_order()` has been expanded.

Each commit is self-contained and passes tests. Happy to split into separate pull requests if it helps.

Closes #276, closes #373.

### CODE REVIEW

File: R/verb-window.R [hunk 1]
```diff
@@ -1,5 +1,24 @@
 #' Override window order and frame
 #'
+#' @description
+#' `window_order()` sets the order for subsequent invocations
+#' of window functions on a lazy table.
+#' The order is active only for the next verb.
```
Comment (hadley) [inline R/verb-window.R:6]: Is that also true for `window_frame()`? 

Maybe it's easier to just leave the `window_order()` persistence as is?

File: R/verb-window.R [hunk 2]
```diff
@@ -51,6 +70,15 @@ sql_build.op_order <- function(op, con, ...) {
 
 # Frame -------------------------------------------------------------------
 
+#' @description
```
Comment (hadley) [inline R/verb-window.R:73]: Can you please put all these docs up with `winder_order()`?

File: tests/testthat/sql/arrange-join.txt [hunk 3]
```diff
@@ -5,10 +5,13 @@
 Message: Joining, by = "a"
 
 <SQL>
+
+Warning: Ignoring sort order.
+Hint: `arrange()` only has an effect if used at the end of a pipe or after `head()`. See `?arrange.tbl_lazy` for details.
```
Comment (hadley) [inline tests/testthat/sql/arrange-join.txt:10]: _before_ head?

File: tests/testthat/sql/arrange-single.txt [hunk 4]
```diff
@@ -35,41 +34,51 @@ ORDER BY `a`
 > # complex mutate
 > lf %>% arrange(a) %>% mutate(a = b) %>% arrange(a)
```
Comment (hadley) [inline tests/testthat/sql/arrange-single.txt:35]: Why do we have so many cases here? They all seem rather similar.

File: R/verb-window.R (hunk 1 above)
Comment (krlmlr) [reply R/verb-window.R:6]: `window_frame()` is different because it never stacks. I agree we should remove it there too: #475.

Comment (krlmlr) [reply R/verb-window.R:6]: If we leave `window_order()` persistence, we need to provide a way to reset it. This also leaves us vulnerable to `. %>% window_order(a) %>% select(-a)`.

I suggest we make a clean cut here, perhaps after the dplyr CRAN release.

Comment (hadley) [reply R/verb-window.R:6]: But clearly `window_order()` and `window_frame()` need to persist the status of the other. That's starting to seem rather complicated to me.

Comment (krlmlr) [reply R/verb-window.R:6]: I don't mind persisting the status of the order in `window_order()` if we provide a way to reset it. What would be a good API -- `window_order(.add = FALSE)` ?

Comment (krlmlr) [reply R/verb-window.R:6]: `window_order()` and `window_frame()` now respect each other, added tests. `window_order()` still overrides the previously set order.

File: tests/testthat/sql/arrange-single.txt (hunk 4 above)
Comment (krlmlr) [reply tests/testthat/sql/arrange-single.txt:35]: Reduced.

**Pull request closed without merging**